ql/src/java/org/apache/hadoop/hive/ql/exec/ColumnStatsUpdateTask.java

```diff
@@ -297,21 +303,34 @@ private ColumnStatisticsDesc getColumnStatsDesc(String dbName,
 
   private int persistColumnStats(Hive db) throws HiveException, MetaException, IOException {
     ColumnStatistics colStats = constructColumnStatsFromInput();
-    ColumnStatisticsDesc colStatsDesc = colStats.getStatsDesc();
-    // We do not support stats replication for a transactional table yet. If we are converting
-    // a non-transactional table to a transactional table during replication, we might get
-    // column statistics but we shouldn't update those.
-    if (work.getColStats() != null &&
-        AcidUtils.isTransactionalTable(getHive().getTable(colStatsDesc.getDbName(),
-                                                          colStatsDesc.getTableName()))) {
-      LOG.debug("Skipped updating column stats for table " +
-                TableName.getDbTable(colStatsDesc.getDbName(), colStatsDesc.getTableName()) +
-                " because it is converted to a transactional table during replication.");
-      return 0;
-    }
-
     SetPartitionsStatsRequest request =
             new SetPartitionsStatsRequest(Collections.singletonList(colStats));
+
+    // Set writeId and validWriteId list for replicated statistics.
+    if (work.getColStats() != null) {
+      String dbName = colStats.getStatsDesc().getDbName();
+      String tblName = colStats.getStatsDesc().getTableName();
+      Table tbl = db.getTable(dbName, tblName);
+      long writeId = work.getWriteId();
+      // If it's a transactional table on source and target, we will get a valid writeId
+      // associated with it. Otherwise it's a non-transactional table on source migrated to a
+      // transactional table on target, we need to craft a valid writeId here.
+      if (AcidUtils.isTransactionalTable(tbl)) {
+        ValidWriteIdList writeIds;
+        if (writeId <= 0) {
```

Comment:
Instead of having this assumption of "writeId <= 0 means migration case", it is better if the caller sets the correct writeId in ColumnStatsUpdateWork itself. 
If it is non-migration case and there is a bug in the caller and passes wrong writeId, then we throw incorrect exception message.

Reply:
We can not set writeId in the ColumnStatsUpdateWork because the writeId for migration is available only after a transaction is opened for migration, which doesn't happen at the load time (when the work is created). Going by the gist of your suggestion, I have set a flag in work to indicate that the writeId should be the one used for migration.